x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/analytics/event/AnalyticsEvent.java

```diff
@@ -0,0 +1,168 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.application.analytics.event;
+
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.xcontent.ParseField;
+import org.elasticsearch.xcontent.ToXContentObject;
+import org.elasticsearch.xcontent.XContentBuilder;
+
+import java.io.IOException;
+import java.util.Locale;
+import java.util.Objects;
+
+/**
+ * This class represents Analytics events object meant to be emitted to the event queue.
+ * Subclasses are implementing the different event types.
+ */
+public abstract class AnalyticsEvent implements Writeable, ToXContentObject {
+
+    public static final ParseField TIMESTAMP_FIELD = new ParseField("@timestamp");
+    public static final ParseField EVENT_FIELD = new ParseField("event");
+    public static final ParseField EVENT_ACTION_FIELD = new ParseField("action");
+    public static final ParseField DATA_STREAM = new ParseField("data_stream");
+    public static final ParseField DATA_STREAM_TYPE_FIELD = new ParseField("type");
+    public static final ParseField DATA_STREAM_NAMESPACE_FIELD = new ParseField("namespace");
+    public static final ParseField DATA_STREAM_DATASET_FIELD = new ParseField("dataset");
+    public static final String DATA_STREAM_TYPE = "behavioral_analytics";
+    public static final String DATA_STREAM_DATASET = "events";
+
+    /**
+     * Analytics context. Used to carry information to parsers.
+     */
+    public interface Context {
+        long eventTime();
+
+        Type eventType();
+
+        String eventCollectionName();
+    }
+
+    /**
+     * Analytics event types.
+     */
+    public enum Type {
+        PAGEVIEW("pageview"),
+        SEARCH("search"),
+        SEARCH_CLICK("search_click");
+
+        private final String typeName;
+
+        Type(String typeName) {
+            this.typeName = typeName;
+        }
+
+        @Override
+        public String toString() {
+            return typeName.toLowerCase(Locale.ROOT);
+        }
+    }
+
+    private final String eventCollectionName;
+
+    private final long eventTime;
+
+    private final AnalyticsEventSessionData session;
+
+    private final AnalyticsEventUserData user;
+
+    protected AnalyticsEvent(String eventCollectionName, long eventTime, AnalyticsEventSessionData session, AnalyticsEventUserData user) {
```

Comment:
Would it make sense to add a version field here, to preserve backward compatibility in future changes?